### PR TITLE
RATEST-283 : NEW E2E Test : RefApp 2.x Form Test

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-283

### CODE REVIEW

**Comment by sherrif10** [comment]
cc @kdaud  .This is now ready to be reviewed

File: .github/workflows/refapp-2.x-forms.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      - name: wait for openmrs instance to start
+        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost:8080/openmrs/login.htm)" != "200" ]]; do sleep 1; done
+      - name: Run qaframework on selenium
+        run: |
+          echo "cucumber.publish.enabled=true" > qaframework-bdd-tests/src/test/resources/cucumber.properties
+          mvn clean install -DskipTests=true
+          cp -f qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties
+          cd qaframework-bdd-tests
+          npm run refapp2Forms
+          curl -sX POST -F messages=@target/cucumber.ndjson https://studio.cucumber.io/cucumber_project/results -H "project-access-token: ${{secrets.CUCUMBER_IO_TOKEN}}" -H "provider: github" -H "repo: ${{github.repository}}" -H "branch: master" -H "revision: ${{github.sha}}"
```
Comment (kdaud) [inline .github/workflows/refapp-2.x-forms.yml:40]: did you just forget to add an empty line at the end of the file to cater for git collaborators?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	@Then("user clicks on delete form")	
+	public void userClicksOnDeleteForm() {
+		homePage.goToManageForm();
+		manageForm.waitForPage();
+		manageForm.deletePath();
+		assertNotNull("Add", manageForm.ADD);
+	}
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:101]: is the empty line needed here?

**Comment by sherrif10** [comment]
Resolved.Thanks.

File: .github/workflows/refapp-2.x-forms.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+name: RefApp 2.x Forms
```
Comment (kdaud) [inline .github/workflows/refapp-2.x-forms.yml:1]: what do you think of `refapp-2.x-forms.yml` vs `refapp-2x-form.yml`?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Feature: Form Management
+
+  Background:
+  Given  a user click on manage html forms
+
+  @selenium
+  @forms
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature:7]: isn't the tag supposed to be `@form`?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import org.openmrs.contrib.qaframework.page.ManageFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageHtmlFormsPage;
+import org.openqa.selenium.By;
+
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:27]: why not align the import order as opposed to the existing files import?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature (hunk 4 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature:7]: There are more than one form in an encounter, Thats the ideal point here.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java (hunk 5 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:27]: > why not align the import order as opposed to the existing files import?

i didnt understand this clearly.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:27]: does taking a look at the one o the existing file [import ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/f8602559fe940272b1d917946dd4879da40f506b/qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FindPatientSteps.java#L12-L23) makes it clear?

**Comment by sherrif10** [comment]
Resolved thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  Scenario: Form management
+   When check the availability of forms
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/15-forms/forms.feature:9]: who checks the availability of the form? is it the system or a user?

**Comment by sherrif10** [comment]
Let me try to explain. What changed. All the folders had to be renamed inorder to make them in order.

**Comment by sherrif10** [comment]
Resolved .Thanks @kdaud

File: README.md [hunk 7]
```diff
@@ -24,14 +24,15 @@ Install & Upgrade Tests |
 [All Chrome Tests](https://ci.openmrs.org/browse/CONTRIB-QA) ![Build Status](https://ci.openmrs.org/plugins/servlet/wittified/build-status/CONTRIB-QA)
 [![All Firefox Tests](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/qa.yml/badge.svg?branch=master)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/qa.yml)
 [![RefApp 2.x Clinical Visit](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-clinical-visit.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-clinical-visit.yml)
-[![RefApp 2.x Patient Condition](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2.x-condition.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2.x-condition.yml)
+[![RefApp 2.x Patient Condition](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-condition.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-condition.yml)
```
Comment (kdaud) [inline README.md:27]: is this change addressing RefApp 2.x Form Test?

**Comment by sherrif10** [comment]
It was an improvement, so I decided to fix it here instead of creating a
separate ticket.

**Comment by kdaud** [comment]
@sherrif10 have you got a chance to take a look at this? https://opensource.com/article/18/6/anatomy-perfect-pull-request

**Comment by sherrif10** [comment]
Resolved Thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature [hunk 8]
```diff
@@ -1 +1,12 @@
-Feature: Forms Management
\ No newline at end of file
+Feature: Form Management
+
+  Background:
+  Given  a user click on manage html forms
+
+  @selenium
+  @forms
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:7]: I see from the selenium implementation, there is one form that is being populated with what a user can do; add a form, edit the form and deleting the form, 

isn't the tag supposed to be `@form`?

**Comment by sherrif10** [comment]
Resolved.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import org.openmrs.contrib.qaframework.page.HomePage;
+import org.openmrs.contrib.qaframework.page.HtmlFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageHtmlFormsPage;
+import org.openqa.selenium.By;
+
+public class FormSteps extends Steps {
+	private static String name = "newFormTest1";
+	private static String description = "description of new form";
+	private static String version = "1.2";
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:32]: isn't these variables supposed to be final?
```js
private static String name = "newFormTest1";
private static String description = "description of new form";
private static String version = "1.2";
```

**Comment by sherrif10** [comment]
Resolved thanks.

**Comment by kdaud** [comment]
is there any reason why [FormTest](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/FormTest.java) is not got rid of within this PR?

**Comment by sherrif10** [comment]
> is there any reason why [FormTest](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/FormTest.java) is not got rid of within this PR?

True that

**Comment by sherrif10** [comment]
Resolved thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature [hunk 10]
```diff
@@ -1 +1,12 @@
-Feature: Forms Management
\ No newline at end of file
+Feature: Form Management
+
+  Background:
+  Given  a user click on manage html forms
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: what happens when the user click on manage html forms?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: > what happens when the user click on manage html forms?

Its the initial point that tracks the page functionality. Let be little detailed.Issue to do with forms. Each single run must first pass through manage html forms to do two things one. First tracks the  avaiability of forms. if the forms is available. then it simply go to home page.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: I expect when a user click on manage htlm form link/app then a certain page is loaded, is this the ideal ?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: Yeah that's the deal still and i it does the same.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: don't you think we need a gherkin step that affirms the page is loaded successfully?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: Added that scenario. at first i wasnt seeing it as a neccessary

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 11]
```diff
@@ -53,6 +53,11 @@ public void tearDown() {
 	public void clickOnManageHtmlFormsPage() {
 		manageHtmlFormsPage = homePage.goToAdministration().clickOnManageHtmlForms();
 	}
+
+	@Then("the system loads manage Html Forms Page")
+	public void systemLoadsManageHtmlFormsPage() {
+		assertPage(manageHtmlFormsPage.waitForPage());
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:59]: I remember we opted to rather assert texts over assert page,  did I get the point right?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature [hunk 12]
```diff
@@ -6,7 +6,8 @@ Feature: Form Management
   @selenium
   @form
   Scenario: Form management
-   When  a user check the availability of forms
+   Then	the system loads manage Html Forms Page
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:9]: I think the better approach is to have this step within the background. @sherrif10 what do you think?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature (hunk 10 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:4]: Resolved thanks.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java (hunk 11 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:59]: Resolved thanks.However i think it depends upon how to use assertPage and AssertTexts

**Comment by sherrif10** [comment]
Resolved thanks for review @kdaud

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+import org.openmrs.contrib.qaframework.page.HtmlFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageHtmlFormsPage;
+import org.openqa.selenium.By;
+
+public class FormSteps extends Steps {
+	private static final String NAME = "newFormTest1";
+	private static final String DESCRIPTION = "description of new form";
+	private static final String VERSION = "1.2";
+	private HomePage homePage;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:34]: do you really need to declare `HomePage` as a variable?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+	private ManageFormsPage manageForm;
+	private ManageHtmlFormsPage manageHtmlFormsPage;
+	private HtmlFormsPage htmlFormsPage;
+	private ClinicianFacingPatientDashboardPage patientDashboardPage;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:38]: do you need to declare ClinicianFacingPatientDashboardPage as a variable?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:38]: We are using its variable in our assertions. Similar to homepage. we are using homePage in the before method

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:38]: why not take advantage of inheritance?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:38]: Are you meaning to use inheritance when testBase doesn't have such class. We simply need to instantiate these classes in before hook and use that advantage of it

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:38]: FormSteps is a child class to Steps. By [inheritance](https://www.javatpoint.com/inheritance-in-java) this makes the child class to access all the properties and behaviors of a parent object(Steps).

By declaring again the variable in the child class creates code redundancy. 

@sherrif10 why not just call the variable name([dashboardPage](https://github.com/openmrs/openmrs-contrib-qaframework/blob/586bd60662dd39fb0bb489e6e8d0b93fae0bdf25/qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/Steps.java#L37)) within FormSteps class?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java]: Resolved thanks.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+import org.openmrs.contrib.qaframework.page.HtmlFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageFormsPage;
+import org.openmrs.contrib.qaframework.page.ManageHtmlFormsPage;
+import org.openqa.selenium.By;
+
+public class FormSteps extends Steps {
+	private static final String NAME = "newFormTest1";
+	private static final String DESCRIPTION = "description of new form";
+	private static final String VERSION = "1.2";
+	private HomePage homePage;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:34]: do you need to declare homePage as a variable?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:34]: Ofcourse am simply using homePage in the before method. Am simply instantiating it to have its instance.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:34]: why not take advantage of inheritance?

**Comment by sherrif10** [comment]
Inheritance doesn't allow instantiating of an object in a class. We must
instantiate an object to use it in before class.

**Comment by kdaud** [comment]
HomePage variable is declared in [ReferenceApplicationTestBase](https://github.com/openmrs/openmrs-contrib-qaframework/blob/586bd60662dd39fb0bb489e6e8d0b93fae0bdf25/qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/helper/ReferenceApplicationTestBase.java#L50) and Steps extends the class.  By inheritance this means all child classes for Steps can access all the properties of the grand parent. In otherwards FormSteps is a grand child to ReferenceApplicationTestBase so you just need to call the variable name(homePage) within the class method  that needs it.

**Comment by sherrif10** [comment]
Resolved thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature [hunk 16]
```diff
@@ -1 +1,13 @@
...
+Feature: Form Management
+
+  Background:
+  Given  a user click on manage html forms
+  Then the system loads manage html forms page
+
+  @selenium
+  @form
+  Scenario: Form management
+   When  a user check the availability of forms
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature:10]:
```suggestion
   When a user check the availability of forms
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/09-form/forms.feature]: Resolved thanks.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+		
+		homePage.go();
+		homePage.goToManageForm();
+		if (!manageForm.addPresent()) {
+			manageForm.delete();
+		}
+	}
+	
+	@Then("a user adds a form in the system")
+	public void addEditDeleteFormTest() throws Exception {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:75]: @sherrif10 take a close look at the method name(`addEditDeleteFormTest()`).

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java]: Ohh my bad. Resolved thanks

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 18]
```diff
@@ -72,7 +70,7 @@ public void instantiateFormPresence() throws InterruptedException {
 	}
 	
 	@Then("a user adds a form in the system")
-	public void addEditDeleteFormTest() throws Exception {
+	public void userClicksOnAddForm() throws Exception {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:73]: what do you think of `userClicksOnAddForm()` vs `addFormTest()`?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+		
+		homePage.go();
+		homePage.goToManageForm();
+		if (!manageForm.addPresent()) {
+			manageForm.delete();
+		}
+	}
+	
+	@Then("a user adds a form in the system")
+	public void userClicksOnAddForm() throws Exception {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/FormSteps.java:73]: what do you think of `userClicksOnAddForm()` vs `addFormTest()`?

**Comment by sherrif10** [comment]
Done.

**Review by kdaud — APPROVED** [review]
Well-done @sherrif10 

cc: @dkayiwa

**Comment by kdaud** [comment]
@sherrif10 do you mind creating some tickets be available on the [QA-KanbanBoard](https://issues.openmrs.org/secure/RapidBoard.jspa?rapidView=240&selectedIssue=RATEST-283) for translating legacy tests into BDD tests ?

**Comment by sherrif10** [comment]
Sure . will create them as soon as possible.

**Pull request merged**